lib/web_ui/lib/src/engine/view_embedder/application_dom/application_dom.dart

```diff
@@ -0,0 +1,95 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../../dom.dart';
+
+import '../hot_restart_cache_handler.dart';
+import 'custom_element_application_dom.dart';
+import 'full_page_application_dom.dart';
+
+/// Provides the API that the FlutterViewEmbedder uses to interact with the DOM.
+///
+/// The base class handles "global" stuff that is shared across implementations,
+/// like handling hot-restart cleanup.
+///
+/// This class is specialized to handle different types of DOM embeddings:
+///
+/// * [FullPageApplicationDom] - The default behavior, where flutter takes
+///   control of the whole web page. This is how Flutter Web used to operate.
+/// * [CustomElementApplicationDom] - Flutter is rendered inside a custom host
+///   element, provided by the web app programmer through the engine
+///   initialization.
+abstract class ApplicationDom {
+  @mustCallSuper
+  ApplicationDom() {
+    // Prepare some global stuff...
+    assert(() {
+      _hotRestartCache = HotRestartCacheHandler();
+      return true;
+    }());
+  }
+
+  factory ApplicationDom.create({DomElement? hostElement}) {
+    if (hostElement != null) {
+      return CustomElementApplicationDom(hostElement);
+    } else {
+      return FullPageApplicationDom();
+    }
+  }
+
+  /// Keeps a list of elements to be cleaned up at hot-restart.
+  HotRestartCacheHandler? _hotRestartCache;
+
+  /// Returns the 'type' (a String for debugging) of the ApplicationDom instance.
+  String get type;
+
+  /// Sets-up the viewport Meta-Tag for the app.
+  void applyViewportMeta();
```

Comment:
Viewport meta is a page-global property. I don't think this class should contain methods that are inherently page-global. Instead, can we have something more generic, such as `void initialize();`, and if an implementation needs to do global things, e.g. `FullPageApplicationDom`, it would do it privately. Others, such as `CustomElementApplicationDom`, wouldn't. For exampe, there would no need for the `// NOOP` implementation of this method.

Reply:
Yeah, at first I extracted a "semantic" API for my own mental mapping, but this can totally be a private method that is used from the public "initialize". I'll simplify the public API surface of the ApplicationDOM classes.